### DISCUSSION TITLE
Storage buffer: multidimensional array

### DISCUSSION THREAD

**Discussion by Alphapage** [gfx-rs/wgpu#5002] [Q&A] [answered]
Hello,

Is it possible to create a multidimensional array like this:
`let mut storage: Vec<Vec<u32>> = Vec::new();`

```
let storage_buffer = device.create_buffer_init(&wgpu::util::BufferInitDescriptor {
            label: None,
            contents: bytemuck::cast_slice(&storage[..]), // don't know how to deal with this here.                   
            usage: wgpu::BufferUsages::STORAGE | wgpu::BufferUsages::COPY_SRC,
        });
```

I would like to get from wgsl:
`var<storage,read> storage: array<array<u32>>;`

Thank you in advance for your answers.

**Comment by kpreid** [comment, marked as answer]
A `Vec<Vec<u32>>` is a pointer to a memory allocation containing pointers to memory allocations containing `u32`s. You can't copy it directly into a GPU buffer for the same reason `bytemuck` doesn't straightforwardly accept it: it's not a single allocation, and it contains pointers which would be meaningless (or at least unsafe) on the GPU. For the same kind of reason, `array<array<u32>>` is not a valid [WGSL array type](https://www.w3.org/TR/WGSL/#array-types).

What you need to do is design a data format which _does_ use a single allocation for the data.

* If the data is rectangular (each inner `Vec` is the same length) then this is simple: use a single `Vec<u32>`, and compute “flattened” indices in it as `x + width * y` — the same way you'd access the texels in a texture. (Rust libraries like `ndarray` can help do this for you.) This is not just for the GPU's sake — the CPU side can also have increased performance by not needing to follow many different pointers to read or write the data, and so `Vec<Vec<T>>` is an inferior data structure for almost *all* rectangular storage applications.
* If the data is not rectangular (“ragged”) then you will still have to use a single vector, but also separately store offsets (in a separate `Vec` and a separate `wgpu::Buffer`) to be able to discover where each row starts and compute the flattened index from that information.

**Comment by Alphapage** [comment]
Yes, I will use this workaround... 
But I hope it will be possible in the future.
Wait and see.
Many thanks.
